### DISCUSSION TITLE
ModernBERT Continued Training Focus 🎯

### DISCUSSION THREAD

**Discussion by Exios66** [LLM-Mailroom-Services/mailroom-issues#203] [Status Reports]
### [Full Repository](https://github.com/LLM-Mailroom-Services/mailroom-ml)

**Diagnosis:** doc_type routing is solved (0.9505 acc, ECE 0.0259) — the bottleneck is the subclass heads under class imbalance, not routing or calibration. `contract` (0.0195 F1) and `correspondence` (0.1289) have collapsed to majority-prior; `corporate_record` (0.2485) and `merger_agreement` (0.2381) are only partially learned. Conditional subclass accuracy (0.5831) is well below the 0.75 P0 target — this is the entire cause of the #112 gate failure.
> 
> **Root cause:** subclass heads aren't getting enough gradient signal relative to the dominant doc_type objective (λ_dt = 0.65), compounded by extreme class imbalance that inverse-frequency weighting (cap 20) isn't fixing on its own.
> 
> **Recommended order:**
> 1. Focal loss (γ=2) on subclass heads — highest leverage, ~1–2 GPU-days
> 2. Two-stage freeze → unfreeze training — ~2 GPU-days
> 3. Rare-class augmentation (contract/correspondence) — only if #1–2 stall
> 4. Hierarchical consistency loss — only if #1–2 stall
> 5. Per-head calibration/threshold retuning — low cost, run anytime
> 
> **Don't:** raise λ_dt, add epochs at the current weighting, or move to ModernBERT-large before the loss signal is fixed — none of these address a representational bottleneck.

### Suggested run order
 
| Priority | Experiment | Est. time | Success criterion |
|---|---|---|---|
| 1 | Focal loss (γ=2), λ_dt=0.65 | ~4.5h | contract F1 ≥ 0.10, correspondence ≥ 0.20 |
| 2 | Freeze→unfreeze + focal loss, λ_dt=0.3 | ~5h | contract ≥ 0.15, correspondence ≥ 0.25, doc_type ≥ 0.94 |
| 3 | Rare-class augmentation + focal loss | ~5h | contract ≥ 0.18, correspondence ≥ 0.28 |
| 4 | Hierarchical consistency (on best run) | ~4.5h | subclass macro-F1 up, no doc_type regression |

> Runs 1–2 are the highest-leverage. If Run 2 can't get `contract` above 0.15, the problem is representational, not optimization — go to Run 3/4 rather than more loss-weight tuning.
> 
> ---
> 
### 1. Per-head focal loss with class-balanced α *(priority 1, ~1–2 GPU-days)*
> Replace weighted cross-entropy on subclass heads with class-weighted focal loss: `L = -α_c(1-p_t)^γ log(p_t)`, γ=2, α_c = N_total/(C·N_c), **per-head** not global α (contract/correspondence likely need higher α given how few positives they have). Run one arm at γ=1 and one at γ=2 to check whether modulation helps or hurts.
> - **Why first:** drop-in loss change, no architecture change, directly targets the collapse. Consistently outperforms weighted CE for rare classes in legal/institutional doc classification literature.
> - **Caution:** a ModernBERT regression paper found reweighted losses can hurt some tasks — watch subclass macro-F1 *and* doc_type accuracy each epoch; revert if doc_type drops >0.5pp.

### 2. Two-stage training: freeze backbone → train heads → unfreeze *(priority 1, ~2 GPU-days)*
> Current `freeze_backbone_epochs = 0` means the backbone updates from step one, and with λ_dt=0.65 the doc_type gradient likely dominates before subclass heads escape the prior basin.
> - **Stage 1** (1 epoch): freeze ModernBERT backbone, train heads only with focal loss, λ_dt=0.
> - **Stage 2** (2 epochs): unfreeze at lower LR (backbone 5e-6, heads 2e-5), reintroduce λ_dt=0.3.
> - **Diagnostic value:** if heads can't separate contract from prior even frozen, the problem is representational; if they can, it's joint-training interference — stage 2 should preserve the separation.
> 
### 3. Head-specific data augmentation for rare subclasses *(priority 2, ~1 GPU-day)*
> Apply only to contract/correspondence positives: back-translation (en→x→en) for structural paraphrases, document-level synonym replacement (EDA), and window-level duplication for multi-window docs (multi-window already outperforms single-window: 0.9592 vs 0.9489, suggesting window-level signal matters).
> - **Risk:** back-translation may introduce label noise on legal text — validate a held-out augmented subset before a full run.

### 4. Hierarchical cross-entropy for subclass heads *(priority 2, ~1 GPU-day)*
> doc_type and subclass form a hierarchy; HXE weights loss across levels so the model is rewarded for the parent level even when it misses the child, giving richer gradient for rare subclasses:
> 
> `L_HXE = -Σ λ^(C^l) log p(C^l | C^(l+1))`
> 
> Cheaper version: add a consistency term coupling the existing heads —
> `L_consistency = KL(p_doc_type ‖ Σ_c p_subclass(c))`, weighted 0.1–0.2.

### 5. Per-head calibration and threshold optimization *(priority 3, ~0.5 GPU-day)*
> `corporate_record`, `correspondence`, and `merger_agreement` are currently excluded from the fast path on ECE (0.071 / 0.1384 / 0.1083). Fit per-head temperature scaling on validation logits (minimize NLL), recompute ECE, and re-include any head that drops below 0.05. Doesn't move macro-F1, but improves deployment utility and may change the selective-risk sweep recommendation.

### What not to do
> - **Don't raise λ_dt** — doc_type is already strong; this only starves the subclass heads further. If anything, drop it to 0.3–0.4 in joint training.
> - **Don't add epochs at current λ_dt** — this is an objective-misalignment problem, not under-training. More epochs just reinforce the prior.
> - **Don't scale to ModernBERT-large yet** — the bottleneck is training signal for rare classes, not capacity. A bigger model may overfit the rare classes worse if the loss isn't fixed first.

**Comment by Exios66** [comment]
### `modernBERT-base` vs. `modernBERT-large`

|                       |ModernBERT-base|ModernBERT-large|
|-----------------------|---------------|----------------|
|Layers                 |22             |28              |
|Hidden size            |768            |1,024           |
|Intermediate (MLP) size|1,152          |2,624           |
|Attention heads        |12             |16              |
|Parameters             |149M           |395M (~2.65x)   |

---

> Both have native support for 8,192 tokens, and both use the same alternating attention scheme — every 3rd layer runs full global attention (RoPE θ=160,000), the other layers run local sliding-window attention with a 128-token window (RoPE θ=10,000). Long-context handling isn't something large adds on top of base; it's baked into the ModernBERT architecture at both sizes. 

---

> The large model has ~2.65x the parameters, concentrated mostly in the MLP blocks (intermediate size more than doubles) plus 6 extra layers. In practice that means:

> - ~2–3x slower per training step at the same batch size, since both FLOPs and sequential depth increase.
> - Higher memory footprint, which often forces a smaller per-device batch size or more gradient accumulation — this compounds the slowdown beyond the raw parameter ratio, especially on memory-constrained GPUs.
> - Large models also typically want a lower backbone LR and can be more sensitive to warmup/schedule than base.

**Comment by Exios66** [comment]
## GRANITE DOCLING

> **HuggingFace**: `ibm-granite/granite-docling-258M`

> **Model size** = `0.3B params`
> **Tensor type** = `BF16`

_"Granite Docling is a multimodal Image-Text-to-Text model engineered for efficient document conversion. It preserves the core features of Docling while maintaining seamless integration with DoclingDocuments to ensure full compatibility..._
> _[...] This model was converted to MLX format from ibm-granite/granite-docling-258M using mlx-vlm version 0.3.3. Refer to the original model card for more details on the model..."_

- [docling-258M](https://huggingface.co/ibm-granite/granite-docling-258M)
- [docling-258M-mlx](https://huggingface.co/ibm-granite/granite-docling-258M-mlx)
> Improved performance on Apple silicon (MLX).

---

### OFFICIAL PUBLICATION

- [Docling: An Efficient Open Source Toolkit...](https://huggingface.co/papers/2501.17887) 
> _IBM Research, Rüschlikon, Switzerland_ ; _(January, 2025)_

---

#### ABSTRACT

```YAML
We introduce Docling, an easy-to-use, self-contained, MIT-
licensed, open-source toolkit for document conversion, that
can parse several types of popular document formats into
a unified, richly structured representation. It is powered by
state-of-the-art specialized AI models for layout analysis
(DocLayNet) and table structure recognition (TableFormer),
and runs efficiently on commodity hardware in a small re-
source budget. Docling is released as a Python package and
can be used as a Python API or as a CLI tool. Docling’s mod-
ular architecture and efficient document representation make
it easy to implement extensions, new features, models, and
customizations. Docling has been already integrated in other
popular open-source frameworks (e.g., LangChain, LlamaIn-
dex, spaCy), making it a natural fit for the processing of doc-
uments and the development of high-end applications. The
open-source community has fully engaged in using, promot-
ing, and developing for Docling, which gathered 10k stars on
GitHub in less than a month and was reported as the No. 1
trending repository in GitHub worldwide in November 2024.
```

---

### CHAT TEMPLATE
> Chat template for ibm-granite/granite-docling-258M-mlx.

> ```jinja 
> {%- for message in messages -%}
> {{- '<|start_of_role|>' + message['role'] + '<|end_of_role|>' -}}
> {%- if message['content'] is string -%}
> {{- message['content'] -}}
> {%- else -%}
> {%- for part in message['content'] -%}
> {%- if part['type'] == 'text' -%}
> {{- part['text'] -}}
> {%- elif part['type'] == 'image' -%}
> {{- '<image>' -}}
> {%- endif -%}
> {%- endfor -%}
> {%- endif -%}
> {{- '<|end_of_text|>
> ' -}}
> {%- endfor -%}
> {%- if add_generation_prompt -%}
> {{- '<|start_of_role|>assistant' -}}
> {%- if controls -%}{{- ' ' + controls | tojson() -}}{%- endif -%}
> {{- '<|end_of_role|>' -}}
> {%- endif -%}
> ```
